text/001x-battery_improvements.md

```diff
@@ -0,0 +1,294 @@
+    </message>
+```
+
+`MAV_BATTERY_STATUS_FLAGS` is a new enum that includes faults, charge state, and battery mode:
+
+```xml
+    <enum name="MAV_BATTERY_STATUS_FLAGS" bitmask="true">
+      <description>Battery status flags for fault, health and state indication.</description>
+      <entry value="1" name="MAV_BATTERY_STATUS_FLAGS_NOT_READY_TO_USE">
+        <description>
+          The battery is not ready to use (fly).
+          Set if the battery has faults or other conditions that make it unsafe to fly with.
+          Note: It will be the logical OR of other status bits (chosen by the manufacturer/integrator).
+        </description>
+      </entry>
+      <entry value="2" name="MAV_BATTERY_STATUS_FLAGS_CHARGING">
+        <description>
+          Battery is charging.
+        </description>
+      </entry>
+      <entry value="4" name="MAV_BATTERY_STATUS_FLAGS_CELL_BALANCING">
+        <description>
+          Battery is cell balancing (during charging).
+          Not ready to use (MAV_BATTERY_STATUS_FLAGS_NOT_READY_TO_USE may be set).
+        </description>
+      </entry>
+      <entry value="8" name="MAV_BATTERY_STATUS_FLAGS_FAULT_CELL_IMBALANCE">
+        <description>
+          Battery cells are not balanced.
+          Not ready to use.
+        </description>
+      </entry>
+      <entry value="16" name="MAV_BATTERY_STATUS_FLAGS_AUTO_DISCHARGING">
+        <description>
+          Battery is auto discharging (towards storage level).
+          Not ready to use (MAV_BATTERY_STATUS_FLAGS_NOT_READY_TO_USE would be set).
+        </description>
+      </entry>
+      <entry value="32" name="MAV_BATTERY_STATUS_FLAGS_REQUIRES_SERVICE">
+        <description>
+          Battery requires service (not safe to fly). 
+          This is set at vendor discretion.
+          It is likely to be set for most faults, and may also be set according to a maintenance schedule (such as age, or number of recharge cycles, etc.).
+        </description>
+      </entry>
+      <entry value="64" name="MAV_BATTERY_STATUS_FLAGS_BAD_BATTERY">
+        <description>
+          Battery is faulty and cannot be repaired (not safe to fly). 
+          This is set at vendor discretion.
+          The battery should be disposed of safely.
+        </description>
+      </entry>
+      <entry value="128" name="MAV_BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED">
+        <description>
+          Automatic battery protection monitoring is enabled. 
+          When enabled, the system will monitor for certain kinds of faults, such as cells being over-voltage.
+          If a fault is triggered then and protections are enabled then a safety fault (MAV_BATTERY_STATUS_FLAGS_FAULT_PROTECTION_SYSTEM) will be set and power from the battery will be stopped.
+          Note that the associated fault (such as MAV_BATTERY_STATUS_FLAGS_FAULT_OVER_VOLT) should always be set whether or not the protection system is engaged.
+        </description>
```

Comment:
Several times people have said "can we separate faults and status flags into separate fields".

A reason to do this is if there is a clear semantic difference between the two that also makes it easier for handling and presentation if they are separate.
Only it is true, do we need to consider how much of a buffer we need for each type, and whether the cost is worth it.
Separating these is pointless if this is not true, and increases the risk that we will run out of bits in the current field "eventually".

The problem is that we do not have a clear definition of what is a "fault vs status", and how we would treat them differently.
If we can define that objectively we can then think about the costs.

The following are IMO definitily statuses.
They reflect modes of operation or information about how other flags are used.
They also aggegate faults and modes to provide a kind of super-test - like STATUS_NOT_READY_TO_USE"

STATUS_FLAGS - DEFINITE
- STATUS_CELL_CHARGING
- STATUS_CELL_BALANCING
- STATUS_NOT_READY_TO_USE
- STATUS_AUTO_DISCHARGING
- STATUS_NOT_READY_TO_USE - combines faults/statuss
- STATUS_PROTECTIONS_ENABLED
- STATUS_CAPACITY_RELATIVE_TO_FULL  
- STATUS_REQUIRES_SERVICE (not faulty, but a hint to user)

The following are IMO definitely faults.
They indicate a _specific_ problem with the battery. You wouldn't take off with these, but you would hope to have them reported when flying and flight stack to take action of some kind (right?)

FAULTS - DEFINITE 
- FAULT_CELL_IMBALANCE
- FAULT_OVER_VOLT  
- FAULT_UNDER_VOLT
- FAULT_OVER_TEMPERATURE
- FAULT_UNDER_TEMPERATURE
- FAULT_FAULT_OVER_CURRENT
- FAULT_SHORT_CIRCUIT

Would you do anything special if you had the ones above in a separate field? i.e. if it isn't useful to separate these then we can stop the discussion.

UNDECIDED - These ones might fit in either.

- STATUS_BAD_BATTERY 
  - unusable battery that should be thrown away.
  - it's beyond the point of flying with or a particular fault being interesting.
  - If we have to split I think this is a status
  
The following aren't faults with the battery itself, but with choosing to use it on this vehicle.
They "feel" more like statuses to me, but it depends how we define fault.
- FAULT_INCOMPATIBLE_VOLTAGE
- FAULT_INCOMPATIBLE_FIRMWARE
- FAULT_INCOMPATIBLE_CELLS_CONFIGURATION

- FAULT_PROTECTION_SYSTEM - fault protection system triggered and power turned off. 
  - Its report that you have no power being delivered.
  - There is nothing an autopilot can do about this really because it has no power.
  - I'm actually not sure why we'd send it since the recipient gets this an is now dead. Perhaps for multi battery systems?
  
Anyway, do we have a specific meaning for fault which makes it useful to separately handle, and can that meaning be applied objectively? @julianoes @AlexKlimaj @dakejahl @hendjoshsr71

Reply:
>can that meaning be applied objectively? 

I think it can only be applied objectively in the context of TI ASICs or other ASICs that have a "fault subsystem." It definitely can't be applied objectively if we want to consider a broader range of "smart" batteries that may or may not have a BMS ASIC. 

I think having them all combined is fine, the only concern is running out of space if other flags need to be added, but hopefully we have done our job defining things well enough that we don't need to add more than an additional 13 flags :sweat_smile:

Reply:
@dakejahl Thanks

> I think it can only be applied objectively in the context of TI ASICs or other ASICs that have a "fault subsystem."...

I disagree - whether or not you support the TI ASICS the faults are objectively faults. They are just faults that some systems are unable to report. I don't see that as different than having a fault that just didn't occur.

> I think having them all combined is fine ...

Me too. If we don't get a different view then it is better to keep them all in one.

Reply:
I think it is fine if they are combined. The ones listed covered most of the cases and we still have space to add more.

Reply:
OK. Agreed.